lib/core/src/Cardano/Wallet/Primitive/Model.hs

```diff
@@ -386,16 +432,47 @@ prefilterBlock b u0 = runState $ do
         => ([(Tx, TxMeta)], UTxO)
         -> Tx
         -> State s ([(Tx, TxMeta)], UTxO)
-    applyTx (!txs, !u) tx = do
-        ourU <- state $ utxoOurs tx
-        let ourIns = Set.fromList (inputs tx) `Set.intersection` dom (u <> ourU)
-        let u' = (u <> ourU) `excluding` ourIns
+    applyTx (!txs, !prevUTxO) tx = do
+        -- The next UTxO state (apply a state transition) (e.g. remove
+        -- transaction outputs we've spent).
+        ourNextUTxO <- filterByAddressM isOurs' (applyTxToUTxO tx prevUTxO)
```

Comment:
I suspect this change might alter the performance characteristics of the `applyTx` function.

Before this change:
- We only apply the `isOurs` filter to the outputs of the transaction we're considering.
- We assume that all entries in the current UTxO set already belong to the wallet, so they don't need to be re-filtered.

After this change:
- We first add all transaction outputs to the UTxO set, regardless of whether they belong to us.
- We then have no option but to apply the `isOurs` filter to all entries in the resulting UTxO set, possibly re-filtering entries that have already been filtered before.

I'm wondering, is there a way to rewrite this section so that we don't need to filter the whole UTxO set on each call to `applyTx`?

Reply:
Given that `applyTxToUTxO tx u = (u <> utxoFromTx tx) "excluding" inputs tx`, we can work algebraically and re-arrange to: `u "excluding" inputs tx <> utxoFromTx tx`, provided that the law `utxoFromTx tx "excluding" inputs tx === utxoFromTx tx` holds. We definitely want to test it holds, because if it doesn't, it's saying that the UTxO set we produce from the transaction isn't in-fact "unspent" at all!

Re-arranging then helps us to redefine `applyTx` as: `u "excluding" inputs tx <> filterByAdressM isOurs' (utxoFromTx tx)`, which is equivalent to the current implementation, under the assumption that `u` only contains transactions that belong to us.

So if my logic is correct, yes I can re-define it to not filter the whole UTxO set.

Reply:
`applyTxToUTxO` is perhaps too course a primitive. I could define a finer primitive `spendTx`, and assert that:
`applyTxToUTxO = spendTx tx u <> utxoFromTx tx`
Then `filterByAddressM isOurs' (applyTxToUTxO tx prevUTxO)`, could be reduced:
```
filterByAddressM isOurs' (applyTxToUTxO tx u)
filterByAddressM isOurs' (spendTx tx u <> utxoFromTx tx) [equivalence mentioned above]
spendTx tx u <> filterByAddressM isOurs' (utxoFromTx tx) [we'd need a new law...]

New law: spendTx tx (filterByAddress f u) = filterByAddress f (spendTx tx u)
```

In my mind, that new law is solid. Spending a tx doesn't change any address properties which could affect a filter.

Reply:
This is something Sandy talks about in "Algebra-Driven Design". Defining a program in terms of it's laws and algebraic properties helps us change performance characteristics while maintaining equivalence.

Reply:
7717363e7b27c0181e94c183720e0d709b522451